controller/gke-cluster-config-handler.go

```diff
@@ -0,0 +1,384 @@
+			return config, err
+		}
+	*/
+	if config.Spec.Imported {
+		config = config.DeepCopy()
+		config.Status.Phase = gkeConfigImportingPhase
+		return h.gkeCC.UpdateStatus(config)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := utils.GetServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	createClusterRequest, err := utils.GenerateGkeClusterCreateRequest(config)
+	if err != nil {
+		return config, err
+	}
+
+	operation, err := svc.Projects.Locations.Clusters.Create(
+		utils.LocationRRN(config.Spec.ProjectID, config.Spec.Region),
+		createClusterRequest).Context(ctx).Do()
+
+	if err != nil && !strings.Contains(err.Error(), "alreadyExists") {
+		return config, err
+	}
+	if err == nil {
+		logrus.Debugf("Cluster %s create is called for project %s and region/zone %s. Status Code %v",
+			config.Spec.ClusterName, config.Spec.ProjectID, config.Spec.Region, operation.HTTPStatusCode)
+	}
+
+	config = config.DeepCopy()
+	config.Status.Phase = gkeConfigCreatingPhase
+	return h.gkeCC.UpdateStatus(config)
+}
+
+func (h *Handler) checkAndUpdate(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+
+	if err := h.validateUpdate(config); err != nil {
+		config = config.DeepCopy()
+		config.Status.Phase = gkeConfigUpdatingPhase
+		var updateErr error
+		config, updateErr = h.gkeCC.UpdateStatus(config)
+		if updateErr != nil {
+			return config, updateErr
+		}
+		return config, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := utils.GetServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	cluster, err := svc.Projects.Locations.Clusters.Get(utils.ClusterRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName)).Context(ctx).Do()
+
+	if cluster.Status == utils.ClusterStatusReconciling {
+		// upstream cluster is already updating, must wait until sending next update
+		logrus.Infof("waiting for cluster [%s] to finish updating", config.Name)
+		if config.Status.Phase != gkeConfigUpdatingPhase {
+			config = config.DeepCopy()
+			config.Status.Phase = gkeConfigUpdatingPhase
+			return h.gkeCC.UpdateStatus(config)
+		}
+		h.gkeEnqueueAfter(config.Namespace, config.Name, 30*time.Second)
+		return config, nil
+	}
+
+	for _, np := range cluster.NodePools {
+
+		if status := np.Status; status == utils.NodePoolStatusReconciling || status == utils.NodePoolStatusStopping ||
+			status == utils.NodePoolStatusProvisioning {
+			if config.Status.Phase != gkeConfigUpdatingPhase {
+				config = config.DeepCopy()
+				config.Status.Phase = gkeConfigUpdatingPhase
+				config, err = h.gkeCC.UpdateStatus(config)
+				if err != nil {
+					return config, err
+				}
+			}
+			logrus.Infof("waiting for cluster [%s] to update nodegroups [%s]", config.Name, np.Name)
+			h.gkeEnqueueAfter(config.Namespace, config.Name, 30*time.Second)
+			return config, nil
```

Comment:
May want to make a helper function for this as it would probably be used a lot.

Reply:
I see you have enqueueUpdate and these particular occurrences use enqueueAfter and enqueueUpdate, could consider making a function for that as well but I understand why you didn't use it now.

Reply:
Still TODO